state-machine.go

```diff
@@ -36,6 +36,11 @@ type HandshakeState interface {
 	Next(hm *HandshakeMessage) (HandshakeState, []HandshakeAction, Alert)
 }
 
+type AppExtensionHandler interface {
+	Send(hs HandshakeType, el *ExtensionList) error
+	Receive(hs HandshakeType, el *ExtensionList) error
+}
```

Comment:
I think a more readable interface would be
```go
Send(HandshakeType) ([]Extension, error)
Receive(HandshakeType, Extension) (error)
```
i.e. `Send` (maybe `Get` or `GetExtensions` would be better) returns a slice of extensions to add, and `Receive` (or `Received`) is called separately for every extension received.